src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractExtensionMethodImportCompletionProvider.cs

```diff
@@ -0,0 +1,195 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion.Log;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal abstract class AbstractExtensionMethodImportCompletionProvider : AbstractImportCompletionProvider
+    {
+        protected abstract bool TryGetReceiverTypeSymbol(SyntaxContext syntaxContext, [NotNullWhen(true)] out ITypeSymbol? receiverTypeSymbol);
+
+        protected override bool ShouldProvideCompletion(Document document, SyntaxContext syntaxContext)
+            => syntaxContext.IsRightOfNameSeparator && IsAddingImportsSupported(document);
+
+        protected async override Task AddCompletionItemsAsync(
+            CompletionContext completionContext,
+            SyntaxContext syntaxContext,
+            HashSet<string> namespaceInScope,
+            bool isExpandedCompletion,
+            CancellationToken cancellationToken)
+        {
+            if (TryGetReceiverTypeSymbol(syntaxContext, out var receiverTypeSymbol))
+            {
+                using var telemetryCounter = new TelemetryCounter();
+                var ticks = Environment.TickCount;
+
+                var project = completionContext.Document.Project;
+
+                using var allTypeNamesDisposer = PooledHashSet<string>.GetInstance(out var allTypeNames);
+                allTypeNames.Add(receiverTypeSymbol.MetadataName);
+                allTypeNames.AddRange(receiverTypeSymbol.GetBaseTypes().Concat(receiverTypeSymbol.GetAllInterfacesIncludingThis()).Select(s => s.MetadataName));
+
+                // interface doesn't inherit from object, but is implicitly convertable to object type.
+                if (receiverTypeSymbol.IsInterfaceType())
+                {
+                    allTypeNames.Add("Object");
+                }
+
+                // We don't want to wait for creating indices from scratch unless user explicitly asked for unimported items (via expander).
+                var matchedMethods = await ExtensionMethodFilteringService.GetPossibleExtensionMethodMatchesAsync(project, allTypeNames.ToImmutableHashSet(), loadOnly: !isExpandedCompletion, cancellationToken).ConfigureAwait(false);
+
+                telemetryCounter.GetFilterTicks = Environment.TickCount - ticks;
+                ticks = Environment.TickCount;
+
+                if (matchedMethods == null)
+                {
+                    telemetryCounter.NoFilter = true;
```

Comment:
what happens to these counters in the event of cancellation in the lines above?  do these counters need to be assigned in finally blocks to ensure the data is meaningful?

Reply:
Changed to only report when not cancelled